File: src/wheel/storage/storage_factory.py
from src.common import Factory
from .storage import Storage
from .storage_models import StorageConfig
from .imp.jsonl_storage import JsonlStorage


class StorageFactory(Factory[Storage]):
    def create(self, config: object) -> Storage:
        if isinstance(config, StorageConfig):
            return JsonlStorage(config.data_dir)
        if isinstance(config, str):
            return JsonlStorage(config)
        return JsonlStorage()

File: lynne-cpp/src/wheel/storage/storage_factory.cpp
#include "wheel/storage/storage_factory.h"
#include "wheel/storage/imp/jsonl_storage.h"

namespace lynne {
namespace wheel {

Storage* StorageFactory::create(const StorageConfig& config) const {
    return new JsonlStorage(config.data_dir);
}

Storage* StorageFactory::create(const char* path) const {
    return new JsonlStorage(std::string(path));
}

Storage* StorageFactory::create() const {
    return new JsonlStorage("data");
}

} // namespace wheel
} // namespace lynne
